File: sdk/python/inkbox/mail/resources/drafts.py
"""Email draft operations."""

from __future__ import annotations

from email.message import Message as HeaderMessage
from typing import TYPE_CHECKING, Any, Iterator
from uuid import UUID

from inkbox.identities.types import _UNSET
from inkbox.mail.types import (
    DraftAttachmentContent,
    DraftDetail,
    DraftRecipients,
    DraftSummary,
    ForwardMode,
    Message,
)

if TYPE_CHECKING:
    from inkbox._http import HttpTransport

_DEFAULT_PAGE_SIZE = 50


class DraftsResource:
    def __init__(self, http: HttpTransport) -> None:
        self._http = http

    def list(
        self, email_address: str, *, page_size: int = _DEFAULT_PAGE_SIZE
    ) -> Iterator[DraftSummary]:
        """Iterate over all drafts in a mailbox, newest first."""
        cursor: str | None = None
        while True:
            page = self._http.get(
                f"/mailboxes/{email_address}/drafts",
                params={"limit": page_size, "cursor": cursor},
            )
            for item in page["items"]:
                yield DraftSummary._from_dict(item)
            if not page["has_more"]:
                break
            cursor = page.get("next_cursor")
            if not cursor:
                break

    def create(
        self,
        email_address: str,
        *,
        to: list[str] | None = None,
        cc: list[str] | None = None,
        bcc: list[str] | None = None,
        subject: str | None = None,
        body_text: str | None = None,
        body_html: str | None = None,
        reply_to: str | None = None,
        thread_id: UUID | str | None = None,
        in_reply_to_message_id: str | None = None,
        references: list[str] | None = None,
        forward_message_id: UUID | str | None = None,
        forward_mode: ForwardMode | str | None = None,
        forward_note_text: str | None = None,
        forward_note_html: str | None = None,
        include_original_attachments: bool = True,
        attachments: list[dict[str, str]] | None = None,
        track_opens: bool = False,
    ) -> DraftDetail:
        """Create an email draft. All message fields may be incomplete."""
        recipients: dict[str, list[str]] = {}
        if to is not None:
            recipients["to"] = to
        if cc is not None:
            recipients["cc"] = cc
        if bcc is not None:
            recipients["bcc"] = bcc
        body: dict[str, Any] = {"recipients": recipients}
        optional = {
            "subject": subject,
            "body_text": body_text,
            "body_html": body_html,
            "reply_to": reply_to,
            "thread_id": thread_id,
            "in_reply_to_message_id": in_reply_to_message_id,
            "references": references,
            "attachments": attachments,
        }
        body.update(
            {
                key: str(value) if isinstance(value, UUID) else value
                for key, value in optional.items()
                if value is not None
            }
        )
        if track_opens:
            body["track_opens"] = True
        if forward_message_id is not None:
            body["forward_message_id"] = str(forward_message_id)
            body["include_original_attachments"] = include_original_attachments
            if forward_mode is not None:
                body["forward_mode"] = ForwardMode(forward_mode).value
            if forward_note_text is not None:
                body["forward_note_text"] = forward_note_text
            if forward_note_html is not None:
                body["forward_note_html"] = forward_note_html
        data = self._http.post(f"/mailboxes/{email_address}/drafts", json=body)
        return DraftDetail._from_dict(data)

    def get(self, email_address: str, draft_id: UUID | str) -> DraftDetail:
        """Get one draft with its bodies and attachment metadata."""
        data = self._http.get(f"/mailboxes/{email_address}/drafts/{draft_id}")
        return DraftDetail._from_dict(data)

    def update(
        self,
        email_address: str,
        draft_id: UUID | str,
        *,
        generation: int,
        recipients: Any = _UNSET,
        subject: Any = _UNSET,
        body_text: Any = _UNSET,
        body_html: Any = _UNSET,
        reply_to: Any = _UNSET,
        thread_id: Any = _UNSET,
        in_reply_to_message_id: Any = _UNSET,
        references: Any = _UNSET,
        track_opens: Any = _UNSET,
        forward_note_text: Any = _UNSET,
        forward_note_html: Any = _UNSET,
    ) -> DraftDetail:
        """Update supplied fields if the draft generation still matches."""
        body: dict[str, Any] = {"generation": generation}
        if recipients is not _UNSET:
            body["recipients"] = (
                recipients.to_wire()
                if isinstance(recipients, DraftRecipients)
                else recipients
            )
        fields = {
            "subject": subject,
            "body_text": body_text,
            "body_html": body_html,
            "reply_to": reply_to,
            "thread_id": thread_id,
            "in_reply_to_message_id": in_reply_to_message_id,
            "references": references,
            "track_opens": track_opens,
            "forward_note_text": forward_note_text,
            "forward_note_html": forward_note_html,
        }
        for key, value in fields.items():
            if value is not _UNSET:
                body[key] = str(value) if isinstance(value, UUID) else value
        data = self._http.patch(
            f"/mailboxes/{email_address}/drafts/{draft_id}", json=body
        )
        return DraftDetail._from_dict(data)

    def delete(
        self, email_address: str, draft_id: UUID | str, *, generation: int
    ) -> None:
        """Delete a draft if its generation still matches."""
        self._http.delete(
            f"/mailboxes/{email_address}/drafts/{draft_id}",
            params={"generation": generation},
        )

    def duplicate(
        self, email_address: str, draft_id: UUID | str, *, generation: int
    ) -> DraftDetail:
        """Create an independent copy of a draft."""
        data = self._http.post(
            f"/mailboxes/{email_address}/drafts/{draft_id}/duplicate",
            json={"generation": generation},
        )
        return DraftDetail._from_dict(data)

    def add_attachments(
        self,
        email_address: str,
        draft_id: UUID | str,
        *,
        generation: int,
        attachments: list[dict[str, str]],
    ) -> DraftDetail:
        """Append attachments to a draft."""
        data = self._http.post(
            f"/mailboxes/{email_address}/drafts/{draft_id}/attachments",
            json={"generation": generation, "attachments": attachments},
        )
        return DraftDetail._from_dict(data)

    def remove_attachment(
        self,
        email_address: str,
        draft_id: UUID | str,
        part_index: int,
        *,
        generation: int,
    ) -> DraftDetail:
        """Remove an attachment from a draft."""
        data = self._http.delete_with_response(
            f"/mailboxes/{email_address}/drafts/{draft_id}/attachments/{part_index}",
            params={"generation": generation},
        )
        return DraftDetail._from_dict(data)

    def download_attachment(
        self,
        email_address: str,
        draft_id: UUID | str,
        part_index: int,
        *,
        generation: int,
    ) -> DraftAttachmentContent:
        """Download attachment bytes from a specific draft generation."""
        response = self._http.get_raw(
            f"/mailboxes/{email_address}/drafts/{draft_id}/attachments/{part_index}",
            accept="application/octet-stream",
            params={"generation": generation},
        )
        header = HeaderMessage()
        header["Content-Disposition"] = response.headers.get(
            "Content-Disposition", ""
        )
        return DraftAttachmentContent(
            content=response.content,
            filename=header.get_filename() or "attachment",
            content_type=response.headers.get(
                "Content-Type", "application/octet-stream"
            ).partition(";")[0],
        )

    def send(
        self, email_address: str, draft_id: UUID | str, *, generation: int
    ) -> Message:
        """Send a draft if its generation still matches."""
        data = self._http.post(
            f"/mailboxes/{email_address}/drafts/{draft_id}/send",
            json={"generation": generation},
        )
        return Message._from_dict(data)

File: sdk/typescript/src/mail/resources/drafts.ts
import { HttpTransport, validateIdempotencyKey } from "../../_http.js";
import {
  DraftAttachmentContent,
  DraftDetail,
  DraftRecipients,
  DraftSummary,
  ForwardMode,
  MailAttachmentInput,
  Message,
  RawCursorPage,
  RawDraftDetail,
  RawDraftSummary,
  RawMessage,
  parseDraftDetail,
  parseDraftSummary,
  parseMessage,
} from "../types.js";

const DEFAULT_PAGE_SIZE = 50;

export interface CreateDraftOptions extends DraftRecipients {
  subject?: string | null;
  bodyText?: string | null;
  bodyHtml?: string | null;
  replyTo?: string | null;
  threadId?: string | null;
  inReplyToMessageId?: string | null;
  references?: string[] | null;
  attachments?: MailAttachmentInput[] | null;
  trackOpens?: boolean;
  forwardMessageId?: string | null;
  forwardMode?: ForwardMode | "inline" | "wrapped";
  includeOriginalAttachments?: boolean;
  forwardNoteText?: string | null;
  forwardNoteHtml?: string | null;
  idempotencyKey?: string;
}

export interface UpdateDraftOptions {
  generation: number;
  recipients?: DraftRecipients | null;
  subject?: string | null;
  bodyText?: string | null;
  bodyHtml?: string | null;
  replyTo?: string | null;
  threadId?: string | null;
  inReplyToMessageId?: string | null;
  references?: string[] | null;
  trackOpens?: boolean;
  forwardNoteText?: string | null;
  forwardNoteHtml?: string | null;
}

function attachmentToWire(attachment: MailAttachmentInput): Record<string, unknown> {
  const wire: Record<string, unknown> = {
    filename: attachment.filename,
    content_type: attachment.contentType,
    content_base64: attachment.contentBase64,
  };
  if (attachment.contentId !== undefined) wire["content_id"] = attachment.contentId;
  return wire;
}

function recipientsToWire(recipients: DraftRecipients): Record<string, unknown> {
  const wire: Record<string, unknown> = {};
  if (recipients.to !== undefined) wire["to"] = recipients.to;
  if (recipients.cc !== undefined) wire["cc"] = recipients.cc;
  if (recipients.bcc !== undefined) wire["bcc"] = recipients.bcc;
  return wire;
}

function setIfDefined(
  body: Record<string, unknown>,
  key: string,
  value: unknown,
): void {
  if (value !== undefined) body[key] = value;
}

function parseFilename(disposition: string | null): string {
  if (!disposition) return "attachment";
  const encoded = disposition.match(/filename\*=UTF-8''([^;]+)/i)?.[1];
  if (encoded) {
    try {
      return decodeURIComponent(encoded);
    } catch {
      return encoded;
    }
  }
  return disposition.match(/filename="([^"]+)"/i)?.[1]
    ?? disposition.match(/filename=([^;]+)/i)?.[1]?.trim()
    ?? "attachment";
}

export class DraftsResource {
  constructor(private readonly http: HttpTransport) {}

  async *list(
    emailAddress: string,
    options?: { pageSize?: number },
  ): AsyncGenerator<DraftSummary> {
    const limit = options?.pageSize ?? DEFAULT_PAGE_SIZE;
    let cursor: string | undefined;
    while (true) {
      const page = await this.http.get<RawCursorPage<RawDraftSummary>>(
        `/mailboxes/${emailAddress}/drafts`,
        { limit, cursor },
      );
      for (const item of page.items) yield parseDraftSummary(item);
      if (!page.has_more) break;
      cursor = page.next_cursor ?? undefined;
      if (!cursor) break;
    }
  }

  async create(emailAddress: string, options: CreateDraftOptions = {}): Promise<DraftDetail> {
    const forwardOnlyKeys = [
      "forwardMode",
      "includeOriginalAttachments",
      "forwardNoteText",
      "forwardNoteHtml",
    ] as const;
    if (
      options.forwardMessageId == null
      && forwardOnlyKeys.some((key) => options[key] !== undefined)
    ) {
      throw new Error("forward options require forwardMessageId");
    }
    if (options.idempotencyKey !== undefined) validateIdempotencyKey(options.idempotencyKey);
    const body: Record<string, unknown> = {};
    if (options.to !== undefined || options.cc !== undefined || options.bcc !== undefined) {
      body["recipients"] = recipientsToWire(options);
    }
    setIfDefined(body, "subject", options.subject);
    setIfDefined(body, "body_text", options.bodyText);
    setIfDefined(body, "body_html", options.bodyHtml);
    setIfDefined(body, "reply_to", options.replyTo);
    setIfDefined(body, "thread_id", options.threadId);
    setIfDefined(body, "in_reply_to_message_id", options.inReplyToMessageId);
    setIfDefined(body, "references", options.references);
    if (options.attachments !== undefined) {
      body["attachments"] = options.attachments?.map(attachmentToWire) ?? null;
    }
    setIfDefined(body, "track_opens", options.trackOpens);
    setIfDefined(body, "forward_message_id", options.forwardMessageId);
    setIfDefined(body, "forward_mode", options.forwardMode);
    setIfDefined(body, "include_original_attachments", options.includeOriginalAttachments);
    setIfDefined(body, "forward_note_text", options.forwardNoteText);
    setIfDefined(body, "forward_note_html", options.forwardNoteHtml);
    const path = `/mailboxes/${emailAddress}/drafts`;
    const created = options.idempotencyKey === undefined
      ? await this.http.post<RawDraftDetail>(path, body)
      : await this.http.post<RawDraftDetail>(path, body, {
        headers: { "Idempotency-Key": options.idempotencyKey },
      });
    return parseDraftDetail(created);
  }

  async get(emailAddress: string, draftId: string): Promise<DraftDetail> {
    return parseDraftDetail(await this.http.get<RawDraftDetail>(
      `/mailboxes/${emailAddress}/drafts/${draftId}`,
    ));
  }

  async update(
    emailAddress: string,
    draftId: string,
    options: UpdateDraftOptions,
  ): Promise<DraftDetail> {
    const body: Record<string, unknown> = { generation: options.generation };
    if (options.recipients !== undefined) {
      body["recipients"] = options.recipients === null
        ? null
        : recipientsToWire(options.recipients);
    }
    setIfDefined(body, "subject", options.subject);
    setIfDefined(body, "body_text", options.bodyText);
    setIfDefined(body, "body_html", options.bodyHtml);
    setIfDefined(body, "reply_to", options.replyTo);
    setIfDefined(body, "thread_id", options.threadId);
    setIfDefined(body, "in_reply_to_message_id", options.inReplyToMessageId);
    setIfDefined(body, "references", options.references);
    setIfDefined(body, "track_opens", options.trackOpens);
    setIfDefined(body, "forward_note_text", options.forwardNoteText);
    setIfDefined(body, "forward_note_html", options.forwardNoteHtml);
    return parseDraftDetail(await this.http.patch<RawDraftDetail>(
      `/mailboxes/${emailAddress}/drafts/${draftId}`, body,
    ));
  }

  async duplicate(emailAddress: string, draftId: string, generation: number): Promise<DraftDetail> {
    return parseDraftDetail(await this.http.post<RawDraftDetail>(
      `/mailboxes/${emailAddress}/drafts/${draftId}/duplicate`, { generation },
    ));
  }

  async delete(emailAddress: string, draftId: string, generation: number): Promise<void> {
    await this.http.delete(`/mailboxes/${emailAddress}/drafts/${draftId}`, {
      params: { generation },
    });
  }

  async addAttachments(
    emailAddress: string,
    draftId: string,
    generation: number,
    attachments: MailAttachmentInput[],
  ): Promise<DraftDetail> {
    return parseDraftDetail(await this.http.post<RawDraftDetail>(
      `/mailboxes/${emailAddress}/drafts/${draftId}/attachments`,
      { generation, attachments: attachments.map(attachmentToWire) },
    ));
  }

  async removeAttachment(
    emailAddress: string,
    draftId: string,
    partIndex: number,
    generation: number,
  ): Promise<DraftDetail> {
    return parseDraftDetail(await this.http.deleteWithResponse<RawDraftDetail>(
      `/mailboxes/${emailAddress}/drafts/${draftId}/attachments/${partIndex}`,
      { params: { generation } },
    ));
  }

  async downloadAttachment(
    emailAddress: string,
    draftId: string,
    partIndex: number,
    generation: number,
  ): Promise<DraftAttachmentContent> {
    const response = await this.http.getBytes(
      `/mailboxes/${emailAddress}/drafts/${draftId}/attachments/${partIndex}`,
      { generation },
    );
    return {
      content: response.data,
      filename: parseFilename(response.headers.get("Content-Disposition")),
      contentType: response.headers.get("Content-Type")?.split(";", 1)[0]?.trim()
        || "application/octet-stream",
    };
  }

  async send(emailAddress: string, draftId: string, generation: number): Promise<Message> {
    return parseMessage(await this.http.post<RawMessage>(
      `/mailboxes/${emailAddress}/drafts/${draftId}/send`, { generation },
    ));
  }
}
